File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n=s.size();
        int i=0,j=0;
        unordered_map<char,int>mp;
        int ans=0;
        while(j<n){
            if(mp[s[j]]>0){
                while(mp[s[j]]>0){
                    mp[s[i]]--;
                    i++;
                }
            }
            mp[s[j]]++;
            ans=max(ans,j-i+1); 
            j++;
        }
        return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer>mp = new HashMap<Character,Integer>();
        int j=0;
        int i=0;
        int ans=0;
        int n=s.length();
        while(j<n){
            if(mp.getOrDefault(s.charAt(j),0)>0){
               while(mp.getOrDefault(s.charAt(j),0)>0){
                   mp.put(s.charAt(i),mp.getOrDefault(s.charAt(i),0)-1);
                   i++;
               } 
            }
            mp.put(s.charAt(j),mp.getOrDefault(s.charAt(j),0)+1);
            ans=Math.max(ans,j-i+1);
            j++;
        }
        return ans;
    }
}